File: content/snippets/cpp/design_patterns_flyweight.cpp
#include <iostream>
#include <memory>
#include <string>
#include <unordered_map>

struct GlyphStyle {
    explicit GlyphStyle(std::string font) : font(std::move(font)) {}
    std::string font;
};

class StyleFactory {
public:
    const GlyphStyle* get(const std::string& font) {
        auto it = styles_.find(font);
        if (it == styles_.end()) {
            it = styles_.emplace(font, std::make_unique<GlyphStyle>(font)).first;
        }
        return it->second.get();
    }

private:
    std::unordered_map<std::string, std::unique_ptr<GlyphStyle>> styles_;
};

int main() {
    StyleFactory factory;
    const GlyphStyle* first = factory.get("mono-12");
    const GlyphStyle* second = factory.get("mono-12");
    std::cout << std::boolalpha << (first == second) << '\n';
}

File: content/snippets/python/design_patterns_flyweight.py
class GlyphStyle:
    def __init__(self, font: str) -> None:
        self.font = font


class StyleFactory:
    def __init__(self) -> None:
        self._styles: dict[str, GlyphStyle] = {}

    def get(self, font: str) -> GlyphStyle:
        if font not in self._styles:
            self._styles[font] = GlyphStyle(font)
        return self._styles[font]


factory = StyleFactory()
first = factory.get("mono-12")
second = factory.get("mono-12")
print(first is second)
